Automattic/jetpack
# Add/WAF ip block list and allow list

#### Changes proposed in this Pull Request:

This PR adds functionality to block or allow requests from specific IP addresses by using the WordPress option (`jetpack_waf_ip_block_list`) in order to block them or a different option (`jetpack_waf_ip_allow_list`) in order to always allow them and skip all rule checking (including the block list).

It also adds code to determine the real user IP by setting trusted proxies and trusted headers (think, load balancers or other reverse proxies) to the `Waf_Request` class before evaluating the request headers to determine the user's real IP address. 

These things will have to be tied together in a follow-up PR in order to actually be used (the user interface does not exist yet to set these things from within the Jetpack WAF settings).

#### Jetpack product discussion



IP block list: 1201069996155224-as-1201817477564453
IP allow list: 1201069996155224-as-1201817477564451
Trusted proxies / headers: pdMJwJ-4X-p2

#### Does this pull request change what data or activity we track or use?



No.

#### Testing instructions:






* Set up jurassic tube for your local dev environment to connect jetpack.
* Checkout this branch.
* Verify the WAF works as it did before.
* Run `wp jetpack-waf setup normal --user=1 --allow-root` in order to re-generate the rules.php (and accompaning new files).
* Add an entry to `wp_options` with option_name = `jetpack_waf_ip_block_list` table like `a:1:{i:0;s:10:"172.18.0.1";}` (exchange for your own IP if different)
  * Regenerate rules files by running the CLI command from above*
  * Verify that all your requests are blocked
* Add an entry to `wp_options` with option_name = `jetpack_waf_ip_allow_list` table like `a:1:{i:0;s:10:"172.18.0.1";}` (exchange for your own IP if different)
  * Regenerate rules files by running the CLI command from above*
  * Verify that all your requests are allowed now
  * Verify that even if you do a request with `?jetpack-waf-always-block=1` query parameter, it is still allowed
* Remove yourself from the allow-list (delete option or change IP)
* Regenerate rules files by running the CLI command from above*
* Verify that if you do a request with `?jetpack-waf-always-block=1` now, you are blocked again

\* = This will be automated later on and will be triggered when the user changes these settings in the WAF user interface.

Feel free to play around with other permutations of these options of course. The allow list outweighs the block list. CIDR blocks are not supported yet (a follow-up PR will do that).
